src/verifier/step4/SparsePolynomial.sol

```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "src/pasta/Vesta.sol";
+import "src/pasta/Pallas.sol";
+
+// Port of Nova' SparsePolynomial (https://github.com/microsoft/Nova/blob/main/src/spartan/polynomial.rs#L121)
+library SparsePolynomialLib {
+    struct Z {
+        uint256 index;
+        uint256 scalar;
+    }
+
+    function evaluate(
+        uint256 num_vars,
+        Z[] memory poly_X,
+        uint256[] memory r_y,
+        uint256 modulus,
+        function (uint256) pure returns (uint256) negateBase
+    ) internal pure returns (uint256) {
+        require(num_vars == r_y.length, "[SparsePolynomialLib.evaluate] num_vars != r_y.length");
+
+        uint256 chi = 1;
+        uint256 z = 0;
+        uint256 r_y_index = 0;
+        uint256 result = 0;
+        for (uint256 j = 0; j < poly_X.length; j++) {
+            z = poly_X[j].index;
+            for (uint256 index = 0; index < num_vars; index++) {
+                if (((z >> num_vars - index - 1) & 1) == 1) {
+                    r_y_index = r_y[index];
+                    assembly {
```

Comment:
Do these 1 or 2 line assembly blocks lower the gas cost? I've avoided them, but I'm curious if it makes a difference.

Reply:
I think using assembly explicitly has some positive effect on gas cost. I tried following `for` loop (inside `evaluate`) without explicit assembly blocks:

```
        for (uint256 j = 0; j < poly_X.length; j++) {
            z = poly_X[j].index;
            for (uint256 index = 0; index < num_vars; index++) {
                if (((z >> num_vars - index - 1) & 1) == 1) {
                    r_y_index = r_y[index];
                    chi = mulmod(chi, r_y_index, modulus);
                } else {
                    r_y_index = negateBase(r_y[index]);
                    chi = mulmod(chi, addmod(1, r_y_index, modulus), modulus);
                }
            }
            // accumulate result
            z = poly_X[j].scalar;
            chi = mulmod(chi, z, modulus);
            result = addmod(result, chi, modulus);
            chi = 1;
        }
``` 

and it gave me following gas cost results:
```
Running 2 tests for test/sparse-poly-evaluation-tests.t.sol:SparseEvaluationTest
[PASS] testSparseEvaluationPrimary() (gas: 29378)
[PASS] testSparseEvaluationSecondary() (gas: 29313)
```

Alternatively, using explicit assembly blocks (as it is currently in this PR) gives [following](https://github.com/lurk-lab/solidity-verifier/actions/runs/5346045851/jobs/9692490464#step:6:57):
```
Running 2 tests for test/sparse-poly-evaluation-tests.t.sol:SparseEvaluationTest
[PASS] testSparseEvaluationPrimary() (gas: 28082)
[PASS] testSparseEvaluationSecondary() (gas: 28017)
```

Reply:
Good to know, thanks for looking into that! It makes me think at some point down the line when we're optimizing we may want to re-look at some of our arithmetic heavy operations again and re-implement them in inline assembly.